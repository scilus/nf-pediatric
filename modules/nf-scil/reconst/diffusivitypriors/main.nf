import nextflow.util.BlankSeparatedList

def compute_noddi_priors ( fa, ad, rd, md, fa_min, fa_max, md_min, roi_radius, prefix, output_directory ) {
    """
    mkdir -p $output_directory

    scil_NODDI_priors.py $fa $ad $rd $md $fa_min $fa_max $md_min $roi_radius \
        --out_txt_1fiber_para $output_directory/${prefix}__para_diff.txt \
        --out_txt_1fiber_perp $output_directory/${prefix}__perp_diff.txt \
        --out_txt_ventricles $output_directory/${prefix}__iso_diff.txt
    """
}


def is_directory ( pathlike ) {
    return !(pathlike instanceof BlankSeparatedList) && pathlike.isDirectory()
}

process RECONST_DIFFUSIVITYPRIORS {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(fa), path(ad), path(rd), path(md), path(priors)

    output:
        tuple val(meta), path("*__para_diff.txt")       , emit: para_diff, optional: true
        tuple val(meta), path("*__perp_diff.txt")       , emit: perp_diff, optional: true
        tuple val(meta), path("*__iso_diff.txt")        , emit: iso_diff, optional: true
        path("priors")                                  , emit: priors, optional: true
        path("mean_para_diff.txt")                      , emit: mean_para_diff, optional: true
        path("mean_perp_diff.txt")                      , emit: mean_perp_diff, optional: true
        path("mean_iso_diff.txt")                       , emit: mean_iso_diff, optional: true
        path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def fa_min = task.ext.fa_min ? "--fa_min " + task.ext.fa_min : ""
    def fa_max = task.ext.fa_max ? "--fa_max " + task.ext.fa_max : ""
    def md_min = task.ext.md_min ? "--md_min " + task.ext.md_min : ""
    def roi_radius = task.ext.roi_radius ? "--roi_radius " + task.ext.roi_radius : ""

    def priors_directory = priors.isEmpty() ? "priors" : !is_directory(priors) ? "priors" : priors
    """
    ${ priors.isEmpty() ? compute_noddi_priors( fa, ad, md, rd, fa_min, fa_max, md_min, roi_radius, prefix, priors_directory ) : ""}

    ${ !priors.isEmpty() && !is_directory(priors) ? "mkdir -p priors && ln $priors priors" : "" }

    cat $priors_directory/*__para_diff.txt > all_para_diff.txt
    awk '{ total += \$1; count++ } END { print total/count }' all_para_diff.txt > mean_para_diff.txt
    cat $priors_directory/*__iso_diff.txt > all_iso_diff.txt
    awk '{ total += \$1; count++ } END { print total/count }' all_iso_diff.txt > mean_iso_diff.txt

    if [[ -e $priors_directory/*__perp_diff.txt ]]
    then
        cat $priors_directory/*__perp_diff.txt > all_perp_diff.txt
        awk '{ total += \$1; count++ } END { print total/count }' all_perp_diff.txt > mean_perp_diff.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_NODDI_priors.py -h

    mkdir priors
    touch priors/${prefix}__para_diff.txt
    touch priors/${prefix}__perp_diff.txt
    touch priors/${prefix}__iso_diff.txt

    touch "mean_para_diff.txt"
    touch "mean_perp_diff.txt"
    touch "mean_iso_diff.txt"


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """
}
