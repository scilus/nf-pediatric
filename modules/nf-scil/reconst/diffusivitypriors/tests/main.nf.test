nextflow_process {

    name "Test Process RECONST_DIFFUSIVITYPRIORS"
    script "../main.nf"
    process "RECONST_DIFFUSIVITYPRIORS"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/diffusivitypriors"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "processing.zip", "light.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("reconst - diffusivitypriors - compute_priors") {
        when {
            process {
            """
            ch_split_test_data = LOAD_DATA.out.test_data_directory
                .branch{
                    processing: it.simpleName == "processing"
                    light: it.simpleName == "light"
                }

             input[0] = ch_split_test_data.processing
                 .map{ test_data_directory -> [
                    [ id:'test', single_end:false ], // meta map
                    file("\${test_data_directory}/fa.nii.gz"),
                    file("\${test_data_directory}/ad.nii.gz"),
                    file("\${test_data_directory}/rd.nii.gz"),
                    file("\${test_data_directory}/md.nii.gz"),
                    []
                ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("reconst - diffusivitypriors - mean priors") {
        when {
            process {
            """
            ch_split_test_data = LOAD_DATA.out.test_data_directory
                .branch{
                    processing: it.simpleName == "processing"
                    light: it.simpleName == "light"
                }

             input[0] = ch_split_test_data.light
                 .map{ test_data_directory -> [
                    [ id:'test', single_end:false ], // meta map
                    [],[],[],[],
                    [file("\${test_data_directory}/priors/s1__iso_diff.txt"),
                    file("\${test_data_directory}/priors/s1__para_diff.txt"),
                    file("\${test_data_directory}/priors/s2__iso_diff.txt"),
                    file("\${test_data_directory}/priors/s2__para_diff.txt")]
            ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
