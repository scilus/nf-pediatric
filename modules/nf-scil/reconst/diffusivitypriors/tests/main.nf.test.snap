{
    "reconst - diffusivitypriors - mean priors": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        "s1__iso_diff.txt:md5,3f13db606b697f920d21efd3c9cf8838",
                        "s1__para_diff.txt:md5,b5ab394820e011a04e3787be3772ad6e",
                        "s2__iso_diff.txt:md5,5b003c147cc5f202d2e8c61acaa5eaea",
                        "s2__para_diff.txt:md5,202cb21086439607b99b81ead3e7f817"
                    ]
                ],
                "4": [
                    "mean_para_diff.txt:md5,4ef4f76ce3a673249bcbaa0eba9ce832"
                ],
                "5": [
                    
                ],
                "6": [
                    "mean_iso_diff.txt:md5,b048050981231e15bfafcda4536e2d05"
                ],
                "7": [
                    "versions.yml:md5,bc73c7d541abf11e86b32e2c596a3dfd"
                ],
                "iso_diff": [
                    
                ],
                "mean_iso_diff": [
                    "mean_iso_diff.txt:md5,b048050981231e15bfafcda4536e2d05"
                ],
                "mean_para_diff": [
                    "mean_para_diff.txt:md5,4ef4f76ce3a673249bcbaa0eba9ce832"
                ],
                "mean_perp_diff": [
                    
                ],
                "para_diff": [
                    
                ],
                "perp_diff": [
                    
                ],
                "priors": [
                    [
                        "s1__iso_diff.txt:md5,3f13db606b697f920d21efd3c9cf8838",
                        "s1__para_diff.txt:md5,b5ab394820e011a04e3787be3772ad6e",
                        "s2__iso_diff.txt:md5,5b003c147cc5f202d2e8c61acaa5eaea",
                        "s2__para_diff.txt:md5,202cb21086439607b99b81ead3e7f817"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bc73c7d541abf11e86b32e2c596a3dfd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-14T19:15:29.241456"
    },
    "reconst - diffusivitypriors - compute_priors": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        "test__iso_diff.txt:md5,3ee78992c128f9b9048549df1cc22c8e",
                        "test__para_diff.txt:md5,8942c91713e155a00bad625911d62f8c",
                        "test__perp_diff.txt:md5,b20dd3f07be6d7317004b929781ae085"
                    ]
                ],
                "4": [
                    "mean_para_diff.txt:md5,8942c91713e155a00bad625911d62f8c"
                ],
                "5": [
                    
                ],
                "6": [
                    "mean_iso_diff.txt:md5,3ee78992c128f9b9048549df1cc22c8e"
                ],
                "7": [
                    "versions.yml:md5,bc73c7d541abf11e86b32e2c596a3dfd"
                ],
                "iso_diff": [
                    
                ],
                "mean_iso_diff": [
                    "mean_iso_diff.txt:md5,3ee78992c128f9b9048549df1cc22c8e"
                ],
                "mean_para_diff": [
                    "mean_para_diff.txt:md5,8942c91713e155a00bad625911d62f8c"
                ],
                "mean_perp_diff": [
                    
                ],
                "para_diff": [
                    
                ],
                "perp_diff": [
                    
                ],
                "priors": [
                    [
                        "test__iso_diff.txt:md5,3ee78992c128f9b9048549df1cc22c8e",
                        "test__para_diff.txt:md5,8942c91713e155a00bad625911d62f8c",
                        "test__perp_diff.txt:md5,b20dd3f07be6d7317004b929781ae085"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bc73c7d541abf11e86b32e2c596a3dfd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-14T19:15:10.632882"
    }
}